### ISSUE TITLE
Emulate interrupts, exceptions, and interrupt controller register behaviour

### ISSUE THREAD

**Issue by eigenform** [hosaka-corp/libstfu#4]
- [x] Handle exceptions blocking further progress (aborts, undef instructions)
- [x] Handle system calls with the undef instruction exception handler
- [ ] Implement IRQs from I/O devices

Interrupts prevent us from executing any further in boot2 (this is a data abort interrupt).
`starlet.c:229:__handle_halt_code(): Died with halt_code=00000015, pc=ffff23a0`

**Comment by eigenform** [comment]
Furthermore, this is confusing because it seems like it happens after setting the MMU/write-buffer  bits in CP15 register 1. Unicorn seems to raise a data abort exception here?

```
...
ffff2394 mrc p15,0x0,r0,cr1,cr0,0x0
ffff2398 orr r0,r0,#0x5
ffff239c mcr p15,0x0,r0,cr1,cr0,0x0
ffff23a0 bx lr
```

**Comment by eigenform** [comment]
This is _apparently_ a consequence of the fact that Unicorn defaults 32-bit big-endian ARM machines to the "cortex-a15" CPU type in their QEMU.

```c
diff --git a/qemu/hw/arm/tosa.c b/qemu/hw/arm/tosa.c
index 4a79e12..12ed00d 100644
--- a/qemu/hw/arm/tosa.c
+++ b/qemu/hw/arm/tosa.c
@@ -22,7 +22,8 @@ static int tosa_init(struct uc_struct *uc, MachineState *machine)
     if (uc->mode & UC_MODE_MCLASS) {
         uc->cpu = (CPUState *)cpu_arm_init(uc, "cortex-m3");
     } else {
-        uc->cpu = (CPUState *)cpu_arm_init(uc, "cortex-a15");
+        uc->cpu = (CPUState *)cpu_arm_init(uc, "arm926");
+        //uc->cpu = (CPUState *)cpu_arm_init(uc, "cortex-a15");
     }
 
     return 0;
```

Just `LD_PRELOAD`'ing my own build with ARM926 seems sufficient to get us past this, and we proceed to die later on due to an unhandled `svc` call (presumably distinct from this data abort exception).

**Comment by eigenform** [comment]
Flags for using ARM926 to-be-addressed upstream in unicorn-engine/unicorn#1165
